No suitable serializer when job-level serialization should happen

When job-level serializer is registered and job is distributed across nodes and serialization should really happen then `com.hazelcast.nio.serialization.HazelcastSerializationException` is thrown. 

This does not happen if:
- given serializer is registered globally
- pipeline is running only on single node

Thrown exception (for mentioned below reproducer):
```
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.mycompany.simple.jet.app.Value'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:173)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:149)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:134)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:122)
	at com.hazelcast.spi.impl.NodeEngineImpl.toData(NodeEngineImpl.java:334)
	at com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl.getPartitionId(InternalPartitionServiceImpl.java:950)
	at com.hazelcast.jet.core.Partitioner$Default.getPartition(Partitioner.java:122)
	at com.hazelcast.jet.core.Edge$KeyPartitioner.getPartition(Edge.java:603)
	at com.hazelcast.jet.impl.execution.OutboundCollector$Partitioned.offer(OutboundCollector.java:193)
	at com.hazelcast.jet.impl.execution.OutboxImpl.doOffer(OutboxImpl.java:251)
	at com.hazelcast.jet.impl.execution.OutboxImpl.offerInternal(OutboxImpl.java:139)
	at com.hazelcast.jet.impl.execution.OutboxImpl.offer(OutboxImpl.java:94)
	at com.hazelcast.jet.core.AbstractProcessor.tryEmit(AbstractProcessor.java:309)
	at com.hazelcast.jet.core.AbstractProcessor.emitFromTraverser(AbstractProcessor.java:408)
	at com.hazelcast.jet.core.AbstractProcessor.emitFromTraverser(AbstractProcessor.java:421)
	at com.hazelcast.jet.impl.processor.GroupP.complete(GroupP.java:88)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.complete(ProcessorTasklet.java:465)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:365)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:240)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366)
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: There is no suitable serializer for class com.mycompany.simple.jet.app.Value
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:507)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:162)
	... 22 more
```

**Steps to reproduce:**
Run following code:
```
Pipeline p = Pipeline.create();
p.readFrom(Sources.<String>list("text"))
        .flatMap(line -> traverseArray(line.toLowerCase().split("\\W+")))
        .map(t -> new Value(t))
        .filter(word -> !word.getValue().isEmpty())
        // if you comment out groupingKey+aggregate stage then it works (it will run on single node)
        .groupingKey(wholeItem())
        .aggregate(counting())
        .writeTo(Sinks.list("list"));


JetConfig config = new JetConfig();
// it works if you register serializer globally
// config.getHazelcastConfig().getSerializationConfig()
//         .addSerializerConfig(new SerializerConfig().setTypeClass(Value.class).setClass(ValueSerializer.class));
JetInstance jet = Jet.newJetInstance(config);
JetInstance jet2 = Jet.newJetInstance(config);

List<String> text = jet.getList("text");
text.add("hello world hello hello world");

JobConfig jc = new JobConfig();
jc.addClass(Main.class, Value.class, ValueSerializer.class);
jc.registerSerializer(Value.class, ValueSerializer.class);
jet.newJob(p, jc).join();
```
where `Value` class is:
```
public class Value {

    private String value;

    public Value() {
    }

    public Value(String value) {
        this.value = value;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Value value1 = (Value) o;
        return value == value1.value;
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
}
```
and `ValueSerializer` is:
```
public class ValueSerializer implements StreamSerializer<Value> {

    @Override
    public int getTypeId() {
        return 1;
    }

    @Override
    public void write(ObjectDataOutput output, Value value) throws IOException {
        output.writeUTF(value.getValue());
    }

    @Override
    public Value read(ObjectDataInput input) throws IOException {
        return new Value(input.readUTF());
    }
}
```

**What you can also try with reproducer:**
- comment out groupingKey+aggregate stage from pipeline - it will not fail (job is not distributed across cluster)
- uncomment serialization config for members - it will not fail because it will register serializer globally
- you can also comment out line `jc.registerSerializer(Value.class, ValueSerializer.class);` to see that it checks that serializer is added but it is not able to actually use it

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Partitioner.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Partitioner.java
@@ -105,6 +105,7 @@ public interface Partitioner<T> extends Serializable {
      * Instances should be retrieved from {@link #defaultPartitioner()}.
      */
     final class Default implements Partitioner<Object> {
+
         private static final long serialVersionUID = 1L;
 
         transient DefaultPartitionStrategy defaultPartitioning;

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java
@@ -148,7 +148,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
         this.nodeEngine = (NodeEngineImpl) nodeEngine;
         this.executionId = executionId;
         initProcSuppliers(jobId, executionId, tempDirectories, serializationService);
-        initDag();
+        initDag(serializationService);
 
         this.ptionArrgmt = new PartitionArrangement(partitionOwners, nodeEngine.getThisAddress());
         JetInstance instance = getJetInstance(nodeEngine);
@@ -332,7 +332,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
         }
     }
 
-    private void initDag() {
+    private void initDag(InternalSerializationService serializationService) {
         final Map<Integer, VertexDef> vMap = vertices.stream().collect(toMap(VertexDef::vertexId, v -> v));
         for (VertexDef v : vertices) {
             v.inboundEdges().forEach(e -> e.initTransientFields(vMap, v, false));
@@ -344,7 +344,8 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                 .flatMap(List::stream)
                 .map(EdgeDef::partitioner)
                 .filter(Objects::nonNull)
-                .forEach(p -> p.init(partitionService::getPartitionId));
+                .forEach(partitioner ->
+                        partitioner.init(object -> partitionService.getPartitionId(serializationService.toData(object))));
     }
 
     private static Collection<? extends Processor> createProcessors(VertexDef vertexDef, int parallelism) {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java
@@ -368,7 +368,8 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
         for (EdgeDef edge : srcVertex.outboundEdges()) {
             Map<Address, ConcurrentConveyor<Object>> memberToSenderConveyorMap = null;
             if (edge.isDistributed()) {
-                memberToSenderConveyorMap = memberToSenderConveyorMap(edgeSenderConveyorMap, edge, jobSerializationService);
+                memberToSenderConveyorMap =
+                        memberToSenderConveyorMap(edgeSenderConveyorMap, edge, jobSerializationService);
             }
             outboundStreams.add(createOutboundEdgeStream(edge, processorIdx, memberToSenderConveyorMap,
                     jobSerializationService));

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java
@@ -144,11 +144,11 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                            long executionId,
                            SnapshotContext snapshotContext,
                            ConcurrentHashMap<String, File> tempDirectories,
-                           InternalSerializationService serializationService) {
+                           InternalSerializationService jobSerializationService) {
         this.nodeEngine = (NodeEngineImpl) nodeEngine;
         this.executionId = executionId;
-        initProcSuppliers(jobId, executionId, tempDirectories, serializationService);
-        initDag(serializationService);
+        initProcSuppliers(jobId, executionId, tempDirectories, jobSerializationService);
+        initDag(jobSerializationService);
 
         this.ptionArrgmt = new PartitionArrangement(partitionOwners, nodeEngine.getThisAddress());
         JetInstance instance = getJetInstance(nodeEngine);
@@ -164,7 +164,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                     new ConcurrentInboundEdgeStream(ssConveyor, 0, 0, true,
                             "ssFrom:" + vertex.name()),
                     new AsyncSnapshotWriterImpl(nodeEngine, snapshotContext, vertex.name(), memberIndex, memberCount,
-                            serializationService),
+                            jobSerializationService),
                     nodeEngine.getLogger(StoreSnapshotTasklet.class.getName() + "."
                             + sanitizeLoggerNamePart(vertex.name())),
                     vertex.name(), higherPriorityVertices.contains(vertex.vertexId()));
@@ -193,13 +193,13 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                         memberIndex,
                         memberCount,
                         tempDirectories,
-                        serializationService
+                        jobSerializationService
                 );
 
                 // createOutboundEdgeStreams() populates localConveyorMap and edgeSenderConveyorMap.
                 // Also populates instance fields: senderMap, receiverMap, tasklets.
                 List<OutboundEdgeStream> outboundStreams = createOutboundEdgeStreams(
-                        vertex, localProcessorIdx, serializationService
+                        vertex, localProcessorIdx, jobSerializationService
                 );
                 List<InboundEdgeStream> inboundStreams = createInboundEdgeStreams(
                         vertex, localProcessorIdx, globalProcessorIndex
@@ -209,7 +209,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
 
                 ProcessorTasklet processorTasklet = new ProcessorTasklet(context,
                         nodeEngine.getExecutionService().getExecutor(TASKLET_INIT_CLOSE_EXECUTOR_NAME),
-                        serializationService, processor, inboundStreams, outboundStreams, snapshotContext,
+                        jobSerializationService, processor, inboundStreams, outboundStreams, snapshotContext,
                         snapshotCollector);
                 tasklets.add(processorTasklet);
                 this.processors.add(processor);
@@ -303,7 +303,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
     private void initProcSuppliers(long jobId,
                                    long executionId,
                                    ConcurrentHashMap<String, File> tempDirectories,
-                                   InternalSerializationService serializationService) {
+                                   InternalSerializationService jobSerializationService) {
         JetService service = nodeEngine.getService(JetService.SERVICE_NAME);
 
         for (VertexDef vertex : vertices) {
@@ -324,7 +324,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                         memberCount,
                         jobConfig.getProcessingGuarantee(),
                         tempDirectories,
-                        serializationService
+                        jobSerializationService
                 ));
             } catch (Exception e) {
                 throw sneakyThrow(e);
@@ -332,7 +332,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
         }
     }
 
-    private void initDag(InternalSerializationService serializationService) {
+    private void initDag(InternalSerializationService jobSerializationService) {
         final Map<Integer, VertexDef> vMap = vertices.stream().collect(toMap(VertexDef::vertexId, v -> v));
         for (VertexDef v : vertices) {
             v.inboundEdges().forEach(e -> e.initTransientFields(vMap, v, false));
@@ -345,7 +345,8 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                 .map(EdgeDef::partitioner)
                 .filter(Objects::nonNull)
                 .forEach(partitioner ->
-                        partitioner.init(object -> partitionService.getPartitionId(serializationService.toData(object))));
+                        partitioner.init(object -> partitionService.getPartitionId(jobSerializationService.toData(object)))
+                );
     }
 
     private static Collection<? extends Processor> createProcessors(VertexDef vertexDef, int parallelism) {
@@ -362,15 +363,15 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
      * Populates {@link #senderMap} and {@link #tasklets} fields.
      */
     private List<OutboundEdgeStream> createOutboundEdgeStreams(VertexDef srcVertex, int processorIdx,
-                                                               InternalSerializationService serializationService) {
+                                                               InternalSerializationService jobSerializationService) {
         final List<OutboundEdgeStream> outboundStreams = new ArrayList<>();
         for (EdgeDef edge : srcVertex.outboundEdges()) {
             Map<Address, ConcurrentConveyor<Object>> memberToSenderConveyorMap = null;
             if (edge.isDistributed()) {
-                memberToSenderConveyorMap = memberToSenderConveyorMap(edgeSenderConveyorMap, edge, serializationService);
+                memberToSenderConveyorMap = memberToSenderConveyorMap(edgeSenderConveyorMap, edge, jobSerializationService);
             }
             outboundStreams.add(createOutboundEdgeStream(edge, processorIdx, memberToSenderConveyorMap,
-                    serializationService));
+                    jobSerializationService));
         }
         return outboundStreams;
     }
@@ -383,7 +384,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
     private Map<Address, ConcurrentConveyor<Object>> memberToSenderConveyorMap(
             Map<String, Map<Address, ConcurrentConveyor<Object>>> edgeSenderConveyorMap,
             EdgeDef edge,
-            InternalSerializationService serializationService
+            InternalSerializationService jobSerializationService
     ) {
         assert edge.isDistributed() : "Edge is not distributed";
         return edgeSenderConveyorMap.computeIfAbsent(edge.edgeId(), x -> {
@@ -397,7 +398,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                 final int destVertexId = edge.destVertex().vertexId();
                 final SenderTasklet t = new SenderTasklet(inboundEdgeStream, nodeEngine, destAddr,
                         destVertexId, edge.getConfig().getPacketSizeLimit(), executionId,
-                        edge.sourceVertex().name(), edge.sourceOrdinal(), serializationService
+                        edge.sourceVertex().name(), edge.sourceOrdinal(), jobSerializationService
                 );
                 senderMap.computeIfAbsent(destVertexId, xx -> new HashMap<>())
                          .computeIfAbsent(edge.destOrdinal(), xx -> new HashMap<>())
@@ -423,10 +424,10 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
     private OutboundEdgeStream createOutboundEdgeStream(EdgeDef edge,
                                                         int processorIndex,
                                                         Map<Address, ConcurrentConveyor<Object>> senderConveyorMap,
-                                                        InternalSerializationService serializationService) {
+                                                        InternalSerializationService jobSerializationService) {
         final int totalPtionCount = nodeEngine.getPartitionService().getPartitionCount();
         OutboundCollector[] outboundCollectors = createOutboundCollectors(
-                edge, processorIndex, senderConveyorMap, serializationService
+                edge, processorIndex, senderConveyorMap, jobSerializationService
         );
         OutboundCollector compositeCollector = compositeCollector(outboundCollectors, edge, totalPtionCount);
         return new OutboundEdgeStream(edge.sourceOrdinal(), compositeCollector);
@@ -435,7 +436,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
     private OutboundCollector[] createOutboundCollectors(EdgeDef edge,
                                                          int processorIndex,
                                                          Map<Address, ConcurrentConveyor<Object>> senderConveyorMap,
-                                                         InternalSerializationService serializationService) {
+                                                         InternalSerializationService jobSerializationService) {
         final int upstreamParallelism = edge.sourceVertex().localParallelism();
         final int downstreamParallelism = edge.destVertex().localParallelism();
         final int numRemoteMembers = ptionArrgmt.remotePartitionAssignment.get().size();
@@ -488,7 +489,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
         // allCollectors[n] where n > 0 is a collector pointing to a remote member _n_.
         final int totalPtionCount = nodeEngine.getPartitionService().getPartitionCount();
         final OutboundCollector[] allCollectors;
-        createIfAbsentReceiverTasklet(edge, ptionsPerProcessor, totalPtionCount, serializationService);
+        createIfAbsentReceiverTasklet(edge, ptionsPerProcessor, totalPtionCount, jobSerializationService);
 
         // assign remote partitions to outbound data collectors
         final Map<Address, int[]> memberToPartitions = ptionArrgmt.remotePartitionAssignment.get();
@@ -511,7 +512,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
     private void createIfAbsentReceiverTasklet(EdgeDef edge,
                                                int[][] ptionsPerProcessor,
                                                int totalPtionCount,
-                                               InternalSerializationService serializationService) {
+                                               InternalSerializationService jobSerializationService) {
         final ConcurrentConveyor<Object>[] localConveyors = localConveyorMap.get(edge.edgeId());
 
         receiverMap.computeIfAbsent(edge.destVertex().vertexId(), x -> new HashMap<>())
@@ -528,7 +529,7 @@ public class ExecutionPlan implements IdentifiedDataSerializable {
                                    ptionsPerProcessor[n]));
                            final OutboundCollector collector = compositeCollector(collectors, edge, totalPtionCount);
                            ReceiverTasklet receiverTasklet = new ReceiverTasklet(
-                                   collector, serializationService,
+                                   collector, jobSerializationService,
                                    edge.getConfig().getReceiveWindowMultiplier(),
                                    getConfig().getInstanceConfig().getFlowControlPeriodMs(),
                                    nodeEngine.getLoggingService(), addr, edge.destOrdinal(), edge.destVertex().name());

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/JobTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/JobTest.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.jet.core;
 
+import com.hazelcast.function.FunctionEx;
 import com.hazelcast.function.SupplierEx;
 import com.hazelcast.jet.JetInstance;
 import com.hazelcast.jet.Job;
@@ -737,7 +738,7 @@ public class JobTest extends SimpleTestInClusterSupport {
         DAG dag = new DAG();
         Vertex source = dag.newVertex("source", () -> new ListSource(new Value(1), new Value(2)));
         Vertex sink = dag.newVertex("sink", DiagnosticProcessors.writeLoggerP());
-        dag.edge(between(source, sink).distributed());
+        dag.edge(between(source, sink).distributed().partitioned(FunctionEx.identity()));
 
         JobConfig config = new JobConfig()
                 .registerSerializer(Value.class, ValueSerializer.class);

